### PR TITLE
Clear XR field managers when upgrading to claim SSA

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

Fixes https://github.com/crossplane/crossplane/issues/5498

@ravlilr this PR achieves the same thing as https://github.com/crossplane/crossplane/pull/5550. I'm not sure whether this approach is better. For now I'm just experimenting.

Take a look at the commentary on the updated `Upgrade` method for details on how it works. The short version is that it removes all field managers. This causes the API server to create a special `before-first-apply` field manager, which it then also deletes (the second time it's called). The result appears to be a successful 'upgrade', with no shared field managers.

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] ~Added or updated unit tests.~
- [x] Added or updated e2e tests.
- [ ] ~Linked a PR or a [docs tracking issue] to [document this change].~
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR.~

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet

### CODE REVIEW

File: internal/controller/apiextensions/claim/syncer_ssa.go
Comment (ravilr) [inline internal/controller/apiextensions/claim/syncer_ssa.go:78]: nit: this isn't entirely accurate as XRs based on Composition functions would have other ssa based managers, right ?

File: internal/controller/apiextensions/claim/syncer_ssa.go
Comment (ravilr) [inline internal/controller/apiextensions/claim/syncer_ssa.go:124]: Does this patch needs to be made conditional on resourceVersion, so as to not overwrite/remove the incorrect  managed fields entry on the server, if they have changed before sending this patch ?

Comment (ravilr) [inline internal/controller/apiextensions/claim/syncer_ssa.go:124]:
```suggestion
		p := []byte(fmt.Sprintf(`[{"op": "remove", "path": "/metadata/managedFields/%d"},
                                          {"op": "replace", "path": "/metadata/resourceVersion", "value": "%s"}]`, idxBFA, obj.GetResourceVersion()))
```

Comment (negz) [reply internal/controller/apiextensions/claim/syncer_ssa.go:124]: I was under the impression that `client.Patch` the supplied `obj` to handle this automatically. Based on what `UpgradeManagedFieldsPatch` was doing, it seems like you're probably right though. Good catch!

https://github.com/kubernetes/client-go/blob/v0.30.0/util/csaupgrade/upgrade.go#L165

Comment (negz) [reply internal/controller/apiextensions/claim/syncer_ssa.go]: Pushed a commit that addresses this.

**Comment by negz** [comment]
@ravilr I've updated the PR based on your review.

@turkenh @phisco @bobh66 - I'll need a review from a maintainer before I can merge.

**Comment by ravilr** [comment]
thanks, LGTM.

**Comment by jbw976** [comment]
This comment from @ravilr about also needing similar clearing of field managers in `composed` resources seems relevant for this PR also: https://github.com/crossplane/crossplane/issues/5498#issuecomment-2071244083

**Comment by negz** [comment]
> This comment from @ravilr about also needing similar clearing of field managers in composed resources seems relevant for this PR also: https://github.com/crossplane/crossplane/issues/5498#issuecomment-2071244083

I'd rather @ravilr or I tackle that in a separate PR, rather than expand the scope of this one.

**Comment by jbw976** [comment]
> I'd rather @ravilr or I tackle that in a separate PR, rather than expand the scope of this one.

Totally reasonable @negz! 🙇‍♂️

**Review by phisco — APPROVED** [review]
just a minor comment, other than that, LGTM! thanks @negz!

File: internal/controller/apiextensions/claim/syncer_ssa.go
Comment (negz) [reply internal/controller/apiextensions/claim/syncer_ssa.go:97]: I think it's useful to have two implementations (Nop and Patching) because it allows us to use the Nop implementation by default, but use the Patching implementation when the claim SSA feature flag is enabled.

Comment (phisco) [reply internal/controller/apiextensions/claim/syncer_ssa.go:97]: Nono, sure, I was talking about the csamanagers argument to the function.

Comment (phisco) [reply internal/controller/apiextensions/claim/syncer_ssa.go:97]: I meant this https://github.com/crossplane/crossplane/pull/5641

**Pull request merged**
